enhancements/20221228-various-storage-class-support.md

```diff
@@ -0,0 +1,98 @@
+
+### User Stories
+
+#### Data archiving or Cold Storage
+
+Users would like to create volume for data archiving. That means the user can specify the StorageClass that only schedule replica to HDD devices. Then the Guest Cluster would create volume associated with this StorageClass from Host Harvester Cluster.
+
+#### Better performance and good data reliability
+
+Sometimes, users want a better performance volume and agree to reduce acceptable data reliability (like lowering the replication from 3 to 2). They can specify the StorageClass that only has two replicas and schedule replication to the SSD/NVME. Then the Guest Cluster can create the volumes that base on this StorageClass and get better performance.
+
+### User Experience In Detail
+
+- Harvester CSI Driver no longer hardcodes the `longhorn` StorageClass name when creating volume from the Guest Cluster. Furthermore, The default StorageClass is associated with the `Default` StorageClass on the host Harvester Cluster.
+
+- Users need to create StorageClass with the parameter `host-storage-class` on the Guest Cluster. This parameter is associated with the StorageClass on the host Harvester Cluster.
+
+### API changes
+
+No
+
+## Design
+
+### Implementation Overview
+
+PVC StorageClass would be set up from `host-storage-class` which is from  StorageClass parameter or controller config host-storage-class.
+The prioritize is StorageClass parameter > controller config.
+If w/o any definition, PVC would use the default StorageClass on the host cluster.
+
+### Notible Variable
+
+- **host-storage-class** - This is associated with the StorageClass Name on the host Harvester Cluster.
+
+### Flow Chart
+
+```mermaid
+flowchart TB
+    Start[Create Volume] -->|Some operations| GetVolParameters(Get Volume Parameters)
+  GetVolParameters --> GetHostStorageClassParam{Check `host-storage-class` on Volume Parameters }
+  GetHostStorageClassParam -->|exist| D[Using `host-storage-class` of volume parameters] --> CreateVolume
+  GetHostStorageClassParam -->|not exist| E[Using the `host-storage-class` from configuration] --> CreateVolume
+  CreateVolume[Create volume and patch StorageClass with `host-storage-class`]
+```
+
+### Test plan
+
+#### Verify the Default StorageClass behavior
+
+  1. Create volum `vol001` with the default StorageClass on the Guest Cluster
+  2. Check the `vol001` should be created with the Default StorageClass on the host Harvester Cluster.
+
+#### Verify the user-defined StorageClass behavior
+
+1. Create a new StorageClass `test` on the host Harvester Cluster.
+2. Create a new StorageClass using following manifest on the Guest Cluster.
+
+    ``` yaml
+    allowVolumeExpansion: true
+    apiVersion: storage.k8s.io/v1
+    kind: StorageClass
+    metadata:
+      name: harvester-SC-test
+    provisioner: driver.harvesterhci.io
+    reclaimPolicy: Delete
+    volumeBindingMode: Immediate
+    parameters:
+      host-storage-class: test
```

Comment:
I'd suggest using `hostStorageClass` to follow the convention.

Reply:
That sounds makes sense. Updated, thanks!